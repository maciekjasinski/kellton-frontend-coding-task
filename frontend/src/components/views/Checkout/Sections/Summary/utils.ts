import { MinifigPartType } from '@/types/minifig';
import axios from 'axios';

export const fetchParts = async (set_id: string) => {
  return await axios
    .get(
      `https://rebrickable.com/api/v3/lego/minifigs/${set_id}/parts/?key=${import.meta.env.VITE_REBRICKABLE_API_KEY}`,
    )
    .then((res) => {
      return res.data.results.map((each: any) => {
        const { part } = each;
        return {
          id: part.part_num,
          name: part.name,
          picture: part.part_img_url || null,
        };
      }) as MinifigPartType[];
    }).catch(() => {
      return [];
    });
};
