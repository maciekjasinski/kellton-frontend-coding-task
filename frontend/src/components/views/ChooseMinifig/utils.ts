import { MinifigType } from '@/types/minifig';
import axios from 'axios';

const getRandomId = (arrayLength: number) => {
  const uniqueIds = new Set<number>();

  while (uniqueIds.size < 3) {
    const id = Math.floor(Math.random() * arrayLength);
    uniqueIds.add(id);
  }

  return Array.from(uniqueIds);
}

export const fetchMinifigs = async () => {
  return await axios
    .get(`https://rebrickable.com/api/v3/lego/minifigs/?key=${import.meta.env.VITE_REBRICKABLE_API_KEY}&page_size=1000&in_theme_id=${import.meta.env.VITE_HARRY_POTTER_THEME_ID}`)
    .then((res) => {
      const randomIds = getRandomId(res.data.count);
      const randomMinifigs = randomIds.map((id: number) => res.data.results[id]);
      return randomMinifigs.map(
        (each: { name: string; set_img_url: string | null; set_url: string; set_num: string }) => {
          return {
            name: each.name,
            picture: each.set_img_url || null,
            detailsUrl: each.set_url,
            setId: each.set_num,
          };
        },
      ) as MinifigType[];
    }).catch(() => {
      return [];
    })
};
