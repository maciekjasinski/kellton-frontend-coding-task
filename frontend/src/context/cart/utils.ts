import { createContext } from 'react';
import { CartContextInterface, State } from './types';

export const CartStateContext = createContext<CartContextInterface | undefined>(undefined);

export const initialValues = {
  selectedMinifig: null,
} as State;
