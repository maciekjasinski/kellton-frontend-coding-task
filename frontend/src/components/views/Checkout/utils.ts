import * as yup from 'yup';
import { parsePhoneNumber } from 'libphonenumber-js';
import { EMAIL_REGEX } from '@/utils';

export const initialValues = {
  name: '',
  surname: '',
  phoneNumber: '',
  email: '',
  dateOfBirth: '',
  address: '',
  city: '',
  state: '',
  zipCode: '',
};

export const validationSchema = yup.object().shape({
  name: yup.string().required('Name is required'),
  surname: yup.string().required('Surname is required'),
  phoneNumber: yup
    .string()
    .test('phoneNumber', 'Phone number is not valid', function () {
      const fieldValue = this.parent.phoneNumber;
      if (fieldValue) {
        try {
          const phoneNumner = parsePhoneNumber(fieldValue, 'US');
          const isValid = phoneNumner.isValid();
          const isPossible = phoneNumner.isPossible();
          return isValid && isPossible;
        } catch (err) {
          return this.createError();
        }
      }
      return this.createError();
    })
    .required('Phone number is required'),
  email: yup.string().required('Email is required').matches(EMAIL_REGEX, 'Invalid email'),
  dateOfBirth: yup
    .date()
    .required('Date of birth is required')
    .test('dateOfBirth', 'Should be greater than 18', function () {
      const value = this.parent.dateOfBirth;
      const dateOfBirth = new Date(value);
      const nowDate = new Date();
      const valid = nowDate.getFullYear() - dateOfBirth.getFullYear() >= 18;
      return !valid ? this.createError() : valid;
    }),
  address: yup.string().required('Address is required'),
  city: yup.string().required('City is required'),
  state: yup.string().required('State is required'),
  zipCode: yup.string().required('Zip code is required'),
});
