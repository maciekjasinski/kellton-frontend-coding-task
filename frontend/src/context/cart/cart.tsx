import { ReactNode, useReducer } from 'react';
import { CartStateContext, initialValues } from './utils';
import { cartReducer } from './reducer';

export const CartProvider = ({ children }: { children: ReactNode }) => {
  const [state, dispatch] = useReducer(cartReducer, initialValues);

  const value = {
    ...state,
    cartDispatch: dispatch,
  };

  return <CartStateContext.Provider value={value}>{children}</CartStateContext.Provider>;
};
