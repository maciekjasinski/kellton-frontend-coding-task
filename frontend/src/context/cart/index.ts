import { useContext } from 'react';

import { CartProvider } from './cart';
import { CartStateContext } from './utils';

const useCart = () => {
  const context = useContext(CartStateContext);
  if (context === undefined) {
    throw new Error('useCart must be used within an CartProvider');
  }
  return context;
};

export { CartProvider, useCart };
