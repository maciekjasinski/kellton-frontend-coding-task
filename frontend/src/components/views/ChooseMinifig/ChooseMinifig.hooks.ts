import { useEffect, useState } from 'react';
import { useCart } from '@/context/cart';
import { MinifigType } from '@/types/minifig';
import { fetchMinifigs } from './utils';
import { useNavigate } from 'react-router-dom';

export const useChooseMinifig = () => {
  const { cartDispatch, selectedMinifig } = useCart();
  const navigate = useNavigate();

  const [minifigs, setMinifigs] = useState<MinifigType[]>([]);
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    cartDispatch({ type: 'clearCart' });
    setIsLoading(true);
    fetchMinifigs()
      .then((res) => {
        setMinifigs(res);
      })
      .finally(() => setIsLoading(false));
  }, []);

  const addToCart = (minifig: MinifigType) => {
    if (minifig.name === selectedMinifig?.name) {
      return cartDispatch({
        type: 'clearCart',
      });
    }
    return cartDispatch({
      type: 'addMinifig',
      minifig: minifig,
    });
  };

  const redirectToCheckout = () => {
    if (selectedMinifig) {
      return navigate('/checkout');
    }
  };

  return {
    minifigs,
    isLoading,
    selectedMinifig,
    addToCart,
    redirectToCheckout,
  };
};
