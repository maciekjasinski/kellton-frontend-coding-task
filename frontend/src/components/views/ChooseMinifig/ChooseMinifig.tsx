import { clsx } from 'clsx';
import { MinifigCard } from '@common/MinifigCard';
import { useChooseMinifig } from './ChooseMinifig.hooks';

export const ChooseMinifig = () => {
  const { minifigs, isLoading, selectedMinifig, addToCart, redirectToCheckout } = useChooseMinifig();
  return (
    <div className="flex min-h-dvh flex-col items-center justify-center gap-8 p-4 lg:gap-20 xl:mx-auto xl:w-maxWidth xl:p-0">
      <h1 className="text-2xl font-extrabold uppercase text-white lg:text-3xl">Choose your minifig</h1>
      <div className="flex w-full flex-col gap-4">
        {isLoading ? (
          <>
            <span className="loading loading-bars loading-lg self-center"></span>
          </>
        ) : minifigs.length ? (
          <div className="flex flex-col gap-4 lg:grid lg:grid-cols-3">
            {minifigs.map((minifig, index: number) => (
              <div
                key={index}
                onClick={() => addToCart(minifig)}
                className={clsx(
                  selectedMinifig?.name === minifig.name ? 'drop-shadow-selectedShadow' : '',
                  'cursor-pointer',
                )}
              >
                <MinifigCard key={index} {...minifig} />
              </div>
            ))}
          </div>
        ) : (
          <div>
            <span>No minifigs found!</span>
          </div>
        )}
      </div>
      <button
        className="btn btn-primary w-full shadow-buttonShadow lg:w-fit"
        disabled={!selectedMinifig}
        onClick={redirectToCheckout}
      >
        Proceed to shipment
      </button>
    </div>
  );
};
