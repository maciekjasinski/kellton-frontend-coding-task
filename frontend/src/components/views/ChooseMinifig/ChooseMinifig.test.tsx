import { expect, describe, it } from 'vitest';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { CartProvider } from '@/context/cart';
import { ChooseMinifig } from './ChooseMinifig';
import { getRandomId } from './utils';

describe('Should generate a list with random ids', () => {
  it('should return an array of length 3', () => {
    expect(getRandomId(10)).toHaveLength(3);
  });

  it('should return an array', () => {
    const result = getRandomId(10);
    expect(result).toBeInstanceOf(Array);
  });

  it('should contain unique ids', () => {
    const result = getRandomId(10);
    const unique = new Set(result);
    expect(unique.size).toBe(result.length);
  });

  it('should return empty array', () => {
    const arrayLength = 0;
    const result = getRandomId(arrayLength);
    expect(result.length).toBe(arrayLength);
  });

  it('should return array with one element', () => {
    const arrayLength = 1;
    const result = getRandomId(arrayLength);
    expect(result.length).toBe(arrayLength);
  });
});

describe('ChooseMinifig', () => {
  it('should create a snapshot', () => {
    const page = renderer.create(
      <BrowserRouter>
        <CartProvider>
          <ChooseMinifig />
        </CartProvider>
      </BrowserRouter>,
    );
    expect(page).toMatchSnapshot();
  });
});
