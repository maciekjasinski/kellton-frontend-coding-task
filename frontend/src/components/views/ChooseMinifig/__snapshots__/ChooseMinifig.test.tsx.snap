// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChooseMinifig > should create a snapshot 1`] = `
<div
  className="flex min-h-dvh flex-col items-center justify-center gap-8 p-4 lg:gap-20 xl:mx-auto xl:w-maxWidth xl:p-0"
>
  <h1
    className="text-2xl font-extrabold uppercase text-white lg:text-3xl"
  >
    Choose your minifig
  </h1>
  <div
    className="flex w-full flex-col gap-4"
  >
    <div>
      <span>
        No minifigs found!
      </span>
    </div>
  </div>
  <button
    className="btn btn-primary w-full shadow-buttonShadow lg:w-fit"
    disabled={true}
    onClick={[Function]}
  >
    Proceed to shipment
  </button>
</div>
`;
